rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+
+GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
```

Comment:
How long has the AWS migration been happening for?  I'm sure it was being talked about (if not begun) when I started, which was almost two years ago now.  However, the AWS migration feels much simpler to me than switching 100% to containers not running on Trusty for all of GOV.UK.

Reply:
The GOV.UK Infrastructure team moved to Reliability Engineering in October of 2017. So longer than that.

Reply:
> However, the AWS migration feels much simpler to me than switching 100% to containers not running on Trusty for all of GOV.UK.

I'm not sure they're entirely comparable. The AWS migration involved data migrations and tricky bits of networking, which I think would be much less difficult in the move to a container platform.

We've learned a lot of lessons from that migration too, so hopefully future work will got a bit faster.

There's also Verify's history of very difficult VMWare -> AWS migrations, which were followed by a relatively quick move to containers. GOV.UK won't be exactly the same, of course.

So the assumption is that this doable in much less than two years, but that's a risky assumption for sure. It would be fairly bad news if we spent a long time on this work, only to realise that we can't complete it before the end of Trusty ESM.